RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+# 0031 Multi Test Run Finalization
+
+# Summary
+This document details a data collector extensibility point to reprocess (combine/merge) attachments obtained across test executions in both design mode and commandline scenarios.
+
+# Motivation
+Today when Test Platform executes tests in parallel only code coverage reports are merged (data collector attachments with uri: `datacollector://microsoft/CodeCoverage/2.0`). For other data collector attachments reprocessing is skipped and all of them are returned by Test Platform.
+
+The [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution. The `dotnet test` command builds the solution and runs a test host application for each test project in the solution. However, currently there is no way to reporcess (combine/merge) data collector attachments associated with each project execution. Code coverage reports are not merged.
+
+When `Run All Tests` is performed in VS, tests for projects can be executed separately based on target platform/target framework amongst other criteria. In this case also combining/merging of data collector attachments is not performed (e.g. code overage reports are not merged). `Analyze Code Coverage for All Tests` is showing coverage report for only some test projects in the run.
+
+# Proposed Changes
+
+Introduce a new `IDataCollectorAttachmentProcessor` interface which can be implemented by Test Platform extensions and provide custom logic to reporcess (combine/merge) data collector attachments. Below interface should be used only for providing logic to reprocess information from data collector attachments from independent test executions. In other words data collector attachment should be modified/created using data from other data collector attachments. This interface should **not be** used for modifying any single data collector attachment. Test Platform will invoke `ProcessAttachmentSets` only if at least 2 data collector attachments related to processor (through `GetExtensionUris`) are created by independent test executions.
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection
+{
+    /// <summary>
+    /// Interface for data collectors add-ins that choose to reprocess generated attachments
+    /// </summary>
+    public interface IDataCollectorAttachmentProcessor
+    {
+        /// <summary>
+        /// Gets the attachments Uris, which are handled by processor
+        /// </summary>
+        IEnumerable<Uri> GetExtensionUris();
+
+        /// <summary>
+        /// Indicates whether ProcessAttachmentSets is associative (e.g. f([a, b, c]) = f([f([a, b]), c]) = f([a, f([b, c])]))
+        /// </summary>
+        bool IsAssociative  { get; }
```

Comment:
should we specifcally use associativity, or perhaps we should just allow for `SupportsPartialMerge`, where for instance `f([a, b, c]) == f(f([a, c]), b)` would also hold.

also, what's the lifetime of these?

perhaps we could have a streaming API where we can pass in ProcessAttachments(...) and then have a MergeAttachments(...) method that does the merge. If a processor can process the attachments on the fly, it would.

Reply:
I've renamed it for SupportsIncrementalProcessing based on Abhitej's comment

I'm not sure if I understand correctly streaming API that you described. We could theoretically have such logic that you call:

FinalizeMultiTestRunAsync([a, b], isLastBatch: false,...); -> not returning any results keeps data in memory in TP, merges a, b
FinalizeMultiTestRunAsync([c, d], isLastBatch: false,...); -> not returning any results keeps data in memory in TP, merges previous results with c, d
FinalizeMultiTestRunAsync([e, f], isLastBatch: false,...); -> not returning any results keeps data in memory in TP, merges previous results with e,f
FinalizeMultiTestRunAsync([g, h], isLastBatch: TRUE,...); -> finally merges previous results with g,h and calls handlers callbacks

Such logic would be the simplest from usage point of view. As client does not need to keep any state on their side. It would complicate stuff inside TP. Interfaces needs to be modified in this case.

Keeping the state (already merged attachments) inside MultiTestRunFinalizationManager would allow to use same logic for parallel tests execution in TP. By this we could improve performance a lot. This would also not require sending back again partial results through translation layer.

from API perspective it could also be:

finalization = await vstestconsolewrapper.StartMultiTestRunFinalizationAsync(handler);
await finalization.ProcessAsync([a,b]); // provides progress info through handler
await  finalization.ProcessAsync([c,d]); // provides progress info through handler
await  finalization.ProcessAsync([e,f]); // provides progress info through handler
await  finalization.CloseAsync();    // provides back results through handler

@drognanar @AbhitejJohn  WDYT?

Reply:
Discussed this with Abhitej. We will stick to current version.